### PR TITLE
Add an integration test for large asset counts

### PR DESCRIPTION
**Pull request by Anviking**
# Issue Number

Relates to #2552 

# Overview

- [x] Add integration test for covering the border in-between where we don't need to split token-bundle change, and where we do need to, to ensure we have the calculation correct

# Comments

```bash
stack ghci cardano-wallet-core-integration cardano-wallet:integration
```

```
λ> import System.Environment
λ> setEnv "NO_POOLS" "1"
λ> :main --match SeaHorse
```

### CODE REVIEW

File: lib/core-integration/src/Test/Integration/Faucet.hs [hunk 1]
```diff
@@ -2239,6 +2250,20 @@ maryIntegrationTestAssets = maMnemonics >>= take 3
         ]
     combined p = simple p `TokenBundle.add` fruit p
 
+    seaHorse is p = bundle p $ flip map is $ \i ->
+        (UnsafeTokenName ("SeaHorse" <> B8.pack (show i)), TokenQuantity 1)
+
+    specialAssets =
+        flip zip (map (\is -> mint (seaHorse is) (maryAssetScripts !! 4)) (chunks 2 [1 .. 200]))
```
Comment (Anviking) [inline lib/core-integration/src/Test/Integration/Faucet.hs]: `maryAssetScripts !! 4` looked like the next un-used one, but should have some nicer way of choosing one.

chunks of 5 assets per address were to big (think we fund 10 or 20 addresses per tx), so went with 2

File: lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs [hunk 2]
```diff
@@ -715,6 +718,41 @@ spec = describe "SHELLEY_TRANSACTIONS" $ do
                 , expectField (#assets . #total . #getApiT) (`shouldNotBe` TokenMap.empty)
                 ]
 
+    forM_ [64 .. 70] $ \n ->
```
Comment (Anviking) [inline lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs]: Would be nice to print / show some size metric too, as these numbers had to be stumbled upon by luck.

File: lib/core/src/Cardano/Wallet.hs [hunk 3]
```diff
@@ -2404,7 +2404,7 @@ instance HasSeverityAnnotation WalletLog where
         MsgTip _ -> Info
         MsgBlocks _ -> Debug
         MsgSelectionStart{} -> Debug
-        MsgSelectionDone{} -> Debug
+        MsgSelectionDone{} -> Notice
```
Comment (Anviking) [inline lib/core/src/Cardano/Wallet.hs]: Was temporary

File: lib/shelley/test/data/cardano-node-shelley/shelley-genesis.yaml [hunk 4]
```diff
@@ -16,9 +16,9 @@ protocolParams:
     major: 0
   minUTxOValue: 1000000
   decentralisationParam: 0.25 # means 75% decentralised
-  maxTxSize: 4096
+  maxTxSize: 40960
   minFeeA: 100
-  maxBlockBodySize: 239857
+  maxBlockBodySize: 2398570
```
Comment (Anviking) [inline lib/shelley/test/data/cardano-node-shelley/shelley-genesis.yaml]: Should revert

File: lib/shelley/test/data/cardano-node-shelley/shelley-genesis.yaml [hunk 5]
```diff
@@ -16,9 +16,9 @@ protocolParams:
     major: 0
   minUTxOValue: 1000000
   decentralisationParam: 0.25 # means 75% decentralised
-  maxTxSize: 4096
+  maxTxSize: 40960
```
Comment (Anviking) [inline lib/shelley/test/data/cardano-node-shelley/shelley-genesis.yaml]: Pretty sure we need raise the maxTxSize, in order to hit the TxOutputTooLarge error, which also occurs at 4096.

Comment (jonathanknowles) [reply lib/shelley/test/data/cardano-node-shelley/shelley-genesis.yaml]: > Pretty sure we need raise the maxTxSize, in order to hit the TxOutputTooLarge error, which also occurs at 4096.

It's my understanding that `maxTxSize` is currently `16384` bytes: https://github.com/input-output-hk/cardano-node/blob/99e678c008405426f4af774f16f4f132539622f5/configuration/cardano/mainnet-shelley-genesis.json#L13

maxTxSize

Comment (Anviking) [reply lib/shelley/test/data/cardano-node-shelley/shelley-genesis.yaml]: 👍 let's use that then

File: lib/core-integration/src/Test/Integration/Faucet.hs (hunk 1 above)
Comment (Anviking) [reply lib/core-integration/src/Test/Integration/Faucet.hs]: Fixed - moved it to a separate definition

File: lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs [hunk 6]
```diff
@@ -715,6 +719,54 @@ spec = describe "SHELLEY_TRANSACTIONS" $ do
...
+                                , "status code "
+                                , show s
+                                ]
+                        (_, Left e) -> Left $ show e
+
+                (nPerAddr,) . verifyRes <$> request @(ApiTransaction n) ctx
+                    (Link.createTransaction @'Shelley wSrc) Default payload
+
+            -- 3. They should all succeed
+            responses `shouldBe` (map ((, Right ()) . snd)  sourceWallets)
```
Comment (Anviking) [inline lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs]: With the incorrect (4096 bytes) max bundle size we get:

Skärmavbild 2021-03-09 kl  15 29 58

I.e. it fails, and the failure is still pretty readable.

**Comment by Anviking** [comment]
bors try

File: lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs [hunk 7]
```diff
@@ -715,6 +719,56 @@ spec = describe "SHELLEY_TRANSACTIONS" $ do
...
+            -- 2. Try spending from each wallet, and record the response.
+            responses <- forM sourceWallets $ \(wSrc, nPerAddr) -> do
+                let seaHorses = map $ \ix ->
+                        (( toText seaHorsePolicyId
+                        , toText $ seaHorseTokenName ix)
+                        , 1)
+                payload <- mkTxPayloadMA @n
+                    destAddr
+                    0
+                    (seaHorses [1, nPerAddr * 2])
```
Comment (Anviking) [inline lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs:751]: Magic `2` here. I did consider removing, but you'd need to do more than `let nAddrs = 2` to have it actually make sense. E.g. choose the relevant indexes in the setup phase. Wasn't sure if worth it.

**Review by rvl — APPROVED** [review]
Nice. In future when we have our own API for minting, asset tests will be better because we can get rid of the asset faucet, and easily mint what's needed in each test case.

**Comment by Anviking** [comment]
bors try

**Comment by Anviking** [comment]
Need to update the two `Transaction with too much metadata` tests after increasing the max tx size...

File: lib/core-integration/src/Test/Integration/Faucet.hs [hunk 8]
```diff
@@ -2239,6 +2242,51 @@ maryIntegrationTestAssets = maMnemonics >>= take 3
         ]
     combined p = simple p `TokenBundle.add` fruit p
 
+-- Assets containing 120 different SeaHorses. Same policyId, but all different.
```
Comment (Anviking) [inline lib/core-integration/src/Test/Integration/Faucet.hs]: Oops, this is out of date now

File: lib/core-integration/src/Test/Integration/Framework/Context.hs [hunk 9]
```diff
@@ -59,6 +61,12 @@ data Context = Context
         -- era-specific assertions.
     , _smashUrl :: Text
         -- ^ Base URL of the mock smash server.
+
+    , _mintSeaHorseAssets :: Int -> [Address] -> IO ()
+        -- ^ TODO: Remove once we can unify cardano-wallet-core-integration and
+        -- cardano-wallet:integration, or when the wallet supports minting.
+        --
+        -- Cannot be used be serveral tests at a time. (!)
```
Comment (jonathanknowles) [inline lib/core-integration/src/Test/Integration/Framework/Context.hs]:
```suggestion
        -- Cannot be used by several tests at a time. (!)
```

File: lib/core-integration/src/Test/Integration/Faucet.hs [hunk 10]
```diff
@@ -2239,6 +2242,51 @@ maryIntegrationTestAssets = maMnemonics >>= take 3
...
+--
+-- This is for testing with large token bundles / asset counts.
+--
+-- This is a separate defenition from maryIntegrationTestAssets to allow minting
```
Comment (jonathanknowles) [inline lib/core-integration/src/Test/Integration/Faucet.hs]:
```suggestion
-- This is a separate definition from maryIntegrationTestAssets to allow minting
```

**Review by jonathanknowles — APPROVED** [review]
Looks great.

Only spotted a couple of typos in comments (see comments.)

**Comment by Anviking** [comment]
🤔 
bors r+

**Pull request merged**